File: app/src/main/java/com/mwigzell/places/data/DataService.java
package com.mwigzell.places.data;


import android.content.Context;
import android.os.AsyncTask;

import com.mwigzell.places.model.Type;
import com.mwigzell.places.redux.ActionCreator;
import com.mwigzell.places.redux.AppAction;
import com.mwigzell.places.redux.AppState;
import com.mwigzell.places.redux.jedux.Store;
import com.mwigzell.places.redux.original.Subscriber;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import timber.log.Timber;

/**
 *
 */
@Singleton
public class DataService implements Subscriber {
    Exception lastError;

    @Inject
    Store<AppAction<Object>, AppState> store;

    @Inject
    ActionCreator actionCreator;

    @Inject
    Context context;

    @Inject
    public DataService() {
        store.subscribe(this);
    }

    public List<Type> loadTypes() throws Exception {
        List<Type> result = new ArrayList<>();
        BufferedReader input = new BufferedReader(new InputStreamReader(context.getAssets().open("types.txt")));
        String line;
        while((line = input.readLine()) != null) {
            result.add(new Type(line));
        }
        return result;
    }

    @Override
    public void onStateChanged() {
        //Timber.d("got state: " + store.getState().state);
        switch(store.getState().state()) {
            case LOAD_TYPES:
                final List<Type> types = store.getState().types();
                AsyncTask loadTypes = new AsyncTask<Object, Void, List<Type>>() {
                    @Override
                    protected List<Type> doInBackground(Object... params) {
                        List<Type> result = null;
                        if (types.size() == 0) {
                            try {
                                result = loadTypes();
                            } catch (Exception e) {
                                lastError = e;
                            }
                        } else {
                            result = types;
                        }
                        return result;
                    }

                    @Override
                    protected void onPostExecute(List<Type> types) {
                        actionCreator.typesLoaded(types);
                    }
                };
                loadTypes.execute();

                break;
        }
    }
}

File: app/src/main/java/com/mwigzell/places/data/DataService.kt
package com.mwigzell.places.data


import android.content.Context
import android.os.AsyncTask

import com.mwigzell.places.model.Type
import com.mwigzell.places.redux.ActionCreator
import com.mwigzell.places.redux.AppAction
import com.mwigzell.places.redux.AppState
import com.mwigzell.places.redux.jedux.Store
import com.mwigzell.places.redux.jedux.Subscriber
import rx.Observable
import rx.Single
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import timber.log.Timber

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.ArrayList

import javax.inject.Inject
import javax.inject.Singleton

/**
 *
 */
@Singleton
class DataService @Inject
constructor(private val context: Context,
            private val actionCreator: ActionCreator,
            private val store: Store<AppAction<Any>, AppState>) : Subscriber {
    lateinit internal var lastError: Exception

    init {
        store.subscribe(this)
    }

    @Throws(Exception::class)
    fun loadTypes(): List<Type> {
        val result = ArrayList<Type>()
        val reader = BufferedReader(InputStreamReader(context!!.assets.open("types.txt")))
        while (true) {
            val line = reader.readLine() ?: break
            result.add(Type(line))
        }
        return result
    }

    fun fetchTypes(): Observable<List<Type>> {
        return Single.fromCallable {
            loadTypes()
        }.toObservable()
    }

    override fun onStateChanged() {
        //Timber.d("got state: " + store.getState().state);
        when (store.state.state()) {
            AppState.States.LOAD_TYPES -> {
                val types = store.state.types()
                if (types.size == 0) {
                    fetchTypes()
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .doOnError {
                                Timber.e(it)}
                            .subscribe {
                                actionCreator.typesLoaded(it)
                            }

                } else {
                    actionCreator.typesLoaded(types)
                }
            }
        }
    }
}
